site/content/blog/2021-04-23-optimizing-go-compiler-2.md

```diff
@@ -0,0 +1,343 @@
+---
+title: "Fixing a bug in the Go compiler as a newbie: a deep dive (II)"
+heading: "Fixing a bug in the Go compiler as a newbie: a deep dive (II)"
+description: ""
+summary: "The second part of this series of posts details the issue we are trying to fix. We are using one of the tools we learned in the first part, Static Single Assignment, to understand each and every line of the intermediate code generated by the compiler. We will even learn a tiny bit of assembly in the way."
+slug: optimizing-go-compiler-2
+date: 2021-04-23T12:01:00-05:00
+categories:
+    - "go"
+author: Alejandro García Montoro
+github: agarciamontoro
+community: alejandro.garcia
+---
+
+This is the second part of a series of posts on the compiler of Go. Make sure you have read [the first part](/blog/optimizing-go-compiler-1) before going on.
+
+In the first part of the series we learned what the issue is and the role of SSA and the rewrite rules in the Go compiler. Let's try and use what we know now to actually investigate the bug.
+
+# The investigation
+
+As we learned in the first part, the issue we need to investigate is that the generated code for the following function is not as optimized as it could be:
+
+```go
+import "encoding/binary"
+
+func copyArrayToSlice(dest []byte, src *[8]byte) {
+	_ = dest[8]                                // 1. bounds check
+	temp := binary.LittleEndian.Uint64(src[:]) // 2. read the contents of the source array into temp
+	binary.LittleEndian.PutUint64(dest, temp)  // 3. write the contents into the destiny slice
+}
+```
+
+Let's use our first tool, SSA, to understand what the compiler is generating. If we generate the SSA for the original function, `copyArrayToSlice`, and look at the `lower` phase, we'll see the following:
+
+```
+b1:
+
+	v1 (?) = InitMem <mem>
+	v7 (5) = Arg <*[8]byte> {src} (b.ptr[*byte], src[*[8]byte])
+	v8 (?) = MOVQconst <int> [8] (b.cap[int], b.len[int])
+	v9 (+6) = Arg <int> {dest} [8] (b.len[int], dest+8[int])
+	v248 (84) = Arg <*byte> {dest}
+	v153 (6) = CMPQconst <flags> [8] v9
+UGT v153 → b43 b3 (likely) (6)
+
+b3: ← b1
+
+	v12 (6) = LoweredPanicBoundsC <mem> [0] v8 v9 v1
+Exit v12 (6)
+
+b24:
+BlockInvalid (+83)
+
+b25:
+BlockInvalid (84)
+
+b27:
+BlockInvalid (85)
+
+b29:
+BlockInvalid (86)
+
+b31:
+BlockInvalid (87)
+
+b33:
+BlockInvalid (88)
+
+b35:
+BlockInvalid (89)
+
+b37:
+BlockInvalid (90)
+
+b39:
+BlockInvalid (91)
```

Comment:
I am wondering if it's really necessary to mention this part. This is just the raw output, and not really helpful in the context of the post. It can be redacted to save some space.

Reply:
I removed all the invalid blocks, but left `b3`, and added a brief explanation on what it does. See https://github.com/mattermost/mattermost-developer-documentation/pull/803/commits/4c3766ee500569280e9b6282ac68aea9c3bc3e8d, thanks!